src/Symfony/Component/Serializer/Encoder/CsvEncoder.php

```diff
@@ -26,6 +26,7 @@ class CsvEncoder implements EncoderInterface, DecoderInterface
     const ENCLOSURE_KEY = 'csv_enclosure';
     const ESCAPE_CHAR_KEY = 'csv_escape_char';
     const KEY_SEPARATOR_KEY = 'csv_key_separator';
+    const ALWAYS_COLLECTION = 'always_collection';
```

Comment:
does convention require us to prefix the value with `csv_`?

Reply:
Good catch @ro0NL, I'm 👍 to use the prefix.

Reply:
Do we wanna toggle the default eventually? Side node; is #25227 related to this same convention in XML? If so should `EncoderInterface` know about the const in general to expose this logic to other encoders?

Reply:
Hi at all of you, thank you @ro0NL to link to my issue.

After some digging in the code, I found that the problem is intentional: there is an explicit check on the number of elements in the collection and if it is only one, the collection is removed.

I don't know why the behavior is this, anyway it is not correct in the scenario that I reported.

So, having an option to make serializer skip the removing of the collection is really helpful if not required: as is, currently I cannot use the Serializer to consume the API (that, just to say, is of PrestaShop, so this problem is related to all people who use Serializer to consume such API.).

The strange thing is that this behavior is not present in the Json Encoder, as I'm consuming also an Json API and never had this same problem.

I've not checked the code of the JsonEncoder, and maybe I will do, but I think that somewhere should be stated clearly which has to be the behavior of ALL encoders when they manage collections.

Becuase currently it is really inconsistent and someone who uses the Serializer without having the patience of going deeper in the code will simply think that it doesn't work (as, in reality, is: it currently doesn't work).

So, in the end, I think that the `ALWAYS_COLLECTION` constant should be a general constant of which all encoders have to be aware and act on.

Reply:
Maybe we should add this on every encoders to be sure that we consistency using the context key

Reply:
BTW I think this one should be focusing on CSV and then ill made a follow PR for all of the others encore, @dunglas @ro0NL @Aerendir WDYT ?

Reply:
I think that for sure this is required on the XML encoder. I don't know if it is required also on the others (`JsonDecode` and `YamlEncoder`).

As this is going to be a global requirement, something that someone MUST take into account when creating a new encoder, I think that this should be fixed in the same PR as we are going to introduce a global constant that will be used by more than only the CSV encoder.

Reply:
Also the `Fixed tickets` property of the opening comment should include also a reference to the issue I opened about the XML encoder... I think...

Reply:
I will take care of the needed modification

Reply:
This is a specific problem to the CSV format. JSON and YAML aren't affected because they have both the notion of array and of object, and don't depend of headers. Regarding the XML encoder, it's not because of the format, it's basically an implementation problem (that will be hard to fix because of BC, but from a broader point of view, the XML support in the Serializer is poor and needs more love).

IMO we should keep this flag CSV only, and after a second though, maybe should be introduce a constructor flag instead of a context option (this behavior must be changed globally IMO).